### ISSUE TITLE
Pull-high-watermark breaks with async source error

### ISSUE THREAD

**Issue by Hilzu** [pull-stream/pull-high-watermark#1]
I have a source that asynchronously fetches new values and might fail (it creates requests to an external API). I'm using `pull-high-watermark` to buffer the results for further processing. If fetching a new value fails and the source calls the callback with an error value, the stream just stalls. I would expect my sink to be called with the error so that I can propagate the error further. Here's a test case that exhibits the problem:

``` javascript
var pull = require('pull-stream')
var HighWaterMark = require('pull-high-watermark')

function source (i=3) {
    return function read (end, cb) {
        // console.log('read', ...arguments)
        if (end) return cb(end)
        setTimeout(() => {
            i = i - 1
            // console.log('i', i)
            if (i === 0) return cb(new Error('bork'))
            if (i < 0) return cb(true)
            cb(null, i)
        }, Math.random() * 3000)
    }
}

function syncSource (i=3) {
    return function read (end, cb) {
        if (end) return cb(end)
        i = i - 1
        if (i === 0) return cb(new Error('bork'))
        if (i < 0) return cb(true)
        cb(null, i)
    }
}

// works, logs the error in the end
// pull(source(), pull.log())

// works when error is thrown sync
// pull(syncSource(), HighWaterMark(), pull.log())

// breaks, just stalls after the error
pull(source(), HighWaterMark(), pull.log())
```

Runnable version: https://tonicdev.com/5767e252aef40512006c247e/pull-high-watermark-async-error

**Comment by dominictarr** [comment]
Thanks. This is now fixed in 0.0.2
and I added you script as a test case.

**Comment by regular** [comment]
should this issue be closed then?
